Make the documentation of Embedded struct clearer (#13560)
vlang/v: doc/docs.md

The convention is not to use receiver names like `self` or `this`,
but a short, preferably one letter long, name.

### Embedded structs

> V supports embedded structs , similar to inherit, parent-subclasses in other programming languages.

Reviewer: It is not similar to inheritance, and the word inheritance is misleading and should not be used.
You are free to feel that way, but it shouldn't be in the documentation.

There is no need to compare it to anything.
But if I had to make an analogy, it would be delegation, not inheritance or parent-child relationships.
It would be more appropriate to compare it to delegation, where the compiler automatically provides wrapper methods for you.

Translated with www.DeepL.com/Translator (free version)

Reviewer: @yuyi98  any comment about this?

Author: In the new version , I write like this

Embedded struct like inherit , struct will automatically have the fields and functions from embedded struct.

Embedded struct is not inherit, there is no parent-child relationship between embedded structs, and can't type cast.

Reviewer 2: I like the new formulation 👍🏻 .

Reviewer 2: except `, and can't type cast.` - that can be just removed

Reviewer 2: oh, sorry - you already changed that 😄 .

Unlike inherit in other programming languages, Embedded struct in v is just [mixin](https://en.wikipedia.org/wiki/Mixin), where there is no parent-child relationship between embedded structs, and can't type cast.

```v

> struct Color {

Reviewer: The current example is simpler and easier to understand. I don't see the need to change it.

Author: I changed the doc

mut:
        r byte
        g byte
        b byte
}

struct Widget {
mut:
        x int
        y int
}

fn (w &Widget) area() int {
        return w.x * w.y
}

struct Button {
        Widget
        Color
        title string
}
```

With embedding Button auto have field and method from `Color` and `Widget` do:

```v oksyntax
mut button := Button{
	title: 'Click me'
        y: 2
}

button.x = 3
assert button.area() == 6
button.r = 255
print(button)
```

output :
```v
Button{
    Widget: Widget{
        x: 3
        y: 2
    }
    Color: Color{
        r: 255
        g: 0
        b: 0
    }
    title: 'Click me'
}
```

> Note: If multiple structures are embedded, fields and functions cannot have same name between them.

Reviewer: It can.

You can call fields and methods of any embedded structure by specifying the structure name.

If no structure name is given, the outermost method of the same name will be called.

The same principle applies to the initialization of fields.

Basically, V embedding should works the same way as Go's embedding.

Reviewer 3: @kahsa  Yes, you're right.

Author: @yuyi98 I fix the doc  see below commit , but I found you can't init a struct like this

```
mut button := Button{
  title: 'Click me'
  Widget{
    x: 4
    y: 5
  }
}
```

The convention is not to use receiver names like `self` or `this`,
but a short, preferably one letter long, name.

### Embedded structs

V support embedded structs .

> Basically, V embedding should work the same way as Golang embedding .

Reviewer 4: The original starting comment was more clear.  The heading says what this section is talking about, so the first new sentence here is redundant.  The second sentence doesn't help anyone who is not familiar with Go embedding.

Author: removed

```v
struct Size {
mut:
        width int
        height int
}

fn (s &Size) area() int {
        return s.width * s.height
}

struct Button {
        Size
        title string
}
```

> With embedding Button auto have field and method from `Size` do:

Reviewer 4: ```suggestion
With embedding, Button automatically has all fields and methods from `Size`, which allows you to do:
```

```v oksyntax
mut button := Button{
	title: 'Click me'

>         height: 2

Reviewer 4: Indentation should be the same as the line above.

}

button.width = 3
assert button.area() == 6
assert button.Size.area() == 6
print(button)
```

output :
```v
Button{
    Size: Size{
        width: 3
        height: 2
    }
    title: 'Click me'
}
```

> Embedded struct like inheritance , struct will automatically have the fields and functions from embedded struct.

Reviewer 4: Redundant sentence, repeating what was stated earlier.

> Embedded struct is not inheritance , there is no parent-child relationship between embedded structs, and can't type cast.

Reviewer: Hmm.. IMO it is subtle to assert that there is no parent-child relationship.
Changing the fields or methods of an embedded structure will naturally change the behaviour of the structure on the embedding side.

I think it's enough to say what embedded struct is that `button.widget.x` can simply be written as `button.x` as if you had defined a wrapper method.
And how it works if there is a method or field with the same name.

I think that's the simplest and most accurate way to describe it.

Reviewer: As a side note, in my own projects I use embedding a lot, but I'm not a big fan of using it in the standard library.
If the standard library is designed from the outset to be inherited, as is the case with OOP languages, then it is carefully designed beforehand to avoid changing the behaviour of the classes later on.
However, the V standard library has not been designed in this way.
So when embedding is used, it is presumed that changes to methods will have a more unpredictable effect.

Author: I edited, as below

Embedded struct like inheritance, struct will automatically have the fields and methods from embedded struct.

Embedded struct is not inheritance, can't type cast between structs and embedded structs.

If you want to access embedded structs, write something like `button.Size`.

Conceptually, embedded struct similar [mixin](https://en.wikipedia.org/wiki/Mixin) in OOP, not base class.

An embedded structs is responsible for implementing a small point and exposing a few methods, just like Lego blocks. 

Don't bother to elaborate a bulky base class with huge methods. Needn't import a forest for a banana.

>  The problem with object-oriented languages is they’ve got all this implicit environment that they carry around with them. You wanted a banana but what you got was a gorilla holding the banana and the entire jungle.
 
—— Joe Armstrong, creator of Erlang progamming language

If multiple embedded structs have functions or fields with the same name, 

> or if functions or fields with the same name are defined in struct, 

Reviewer 4: ```suggestion
or if functions or fields with the same name are defined in the struct, 
```

then call or assign embedded struct like `button.Size.area()`.

> You can also initialize an embedded struct like :

Reviewer 4: ```suggestion
You can also initialize an embedded struct:
```

```v
mut button := Button{
    Size: Size{
        x: 4
        y: 5
    }
}
```

> or assign like :

Reviewer 4: ```suggestion
or assign values:
```

}
```

> Embedded struct like inheritance, struct will automatically have the fields and functions from embedded struct.

Reviewer 4: ```suggestion
Similar to inheritance, structs will automatically have the fields and functions from embedded structs.
```

> Embedded struct is not inheritance, can't type cast between structs and embedded structs.

Reviewer 4: ```suggestion
Unlike inheritance, you cannot type cast between structs and embedded structs.
```

> If you want to access embedded structs, write something like `button.Size`.

Reviewer 4: ```suggestion
If you need to access embedded structs directly, use an explicit reference like `button.Size`.
```

> Conceptually, embedded struct similar [mixin](https://en.wikipedia.org/wiki/Mixin) in OOP, not base class.

Reviewer 4: ```suggestion
Conceptually, embedded structs are similar to [mixin](https://en.wikipedia.org/wiki/Mixin)s in OOP, not base classes.
```

> An embedded structs is responsible for implementing a small point and exposing a few functions, just like Lego blocks. 

Reviewer 4: ```suggestion
An embedded structs is responsible for implementing a common structure and exposing a few functions, just like Lego blocks.
```

> Don't bother to elaborate a bulky base class with huge functions. Needn't import a forest for a banana.

Reviewer 4: ```suggestion
It is not recommended to create a bulky base class with a huge number of fields or functions. There is no need to import a forest for a banana.
```
